src/main/java/io/aiven/kafka/connect/gcs/TopicPartitionRecordGrouper.java

```diff
@@ -85,39 +77,64 @@ public void put(final SinkRecord record) {
 
         final TopicPartition tp = new TopicPartition(record.topic(), record.kafkaPartition());
         final SinkRecord currentHeadRecord = currentHeadRecords.computeIfAbsent(tp, ignored -> record);
-        final String filename = renderFilename(tp, currentHeadRecord);
+        final RecordKey recordKey = generateRecordKey(tp, currentHeadRecord);
 
-        if (shouldCreateNewFile(filename)) {
+        if (shouldCreateNewFile(recordKey)) {
             // Create new file using this record as the head record.
             currentHeadRecords.put(tp, record);
-            final String newFilename = renderFilename(tp, record);
-            fileBuffers.computeIfAbsent(newFilename, ignored -> new ArrayList<>()).add(record);
+            final RecordKey newRecordKey = generateRecordKey(tp, record);
+            fileBuffers.computeIfAbsent(newRecordKey, ignored -> new ArrayList<>()).add(record);
         } else {
-            fileBuffers.computeIfAbsent(filename, ignored -> new ArrayList<>()).add(record);
+            fileBuffers.computeIfAbsent(recordKey, ignored -> new ArrayList<>()).add(record);
         }
     }
 
-    private String renderFilename(final TopicPartition tp, final SinkRecord headRecord) {
+    private RecordKey generateRecordKey(final TopicPartition tp, final SinkRecord headRecord) {
+        final Function<Parameter, String> setKafkaOffset =
+            usePaddingParameter -> usePaddingParameter.asBoolean()
+                ? String.format("%020d", headRecord.kafkaOffset())
+                : Long.toString(headRecord.kafkaOffset());
+
+        final Function<Parameter, String> setTimestamp =
+            tsParameter -> !tsParameter.isEmpty()
+                ? tsSource.time().format(DateTimeFormatter.ofPattern(tsParameter.value()))
+                : Long.toString(tsSource.time().toInstant().getEpochSecond());
```

Comment:
Do I understand right, that this will substitute epoch seconds when `unit` is not present?

Reply:
yes you are right

Reply:
I have some doubts it's the way to go: substituting epoch seconds here is not documented and not expected to happen. Is it possible to enforce the `unit` presence on the validation stage?

Reply:
as we discussed i excluded unix time and make `unit` parameter mandatory for the `timestamp` variable